src/VisualStudio/Core/Def/Implementation/ProjectSystem/AbstractOpenFileTracker.cs

```diff
@@ -0,0 +1,314 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.CodeAnalysis.Shared.TestHooks;
+using Microsoft.VisualStudio.ComponentModelHost;
+using Microsoft.VisualStudio.Editor;
+using Microsoft.VisualStudio.Shell.Interop;
+using Task = System.Threading.Tasks.Task;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem
+{
+    /// <summary>
+    /// Subscribes to the running document table and connects/disconnects files to files that are opened.
+    /// </summary>
+    internal abstract class AbstractOpenFileTracker<TWorkspace> where TWorkspace : Workspace
+    {
+        protected readonly ForegroundThreadAffinitizedObject ForegroundAffinitization;
+        protected readonly TWorkspace Workspace;
+        protected readonly IVsRunningDocumentTable4 RunningDocumentTable;
```

Comment:
Is the expectation that other derives _outside_ of Roslyn will use this? My worry is whether we're best off making these private and expect the derived type to re-query if it needs them; otherwise we're exposing more here than we may want.